File: avs/internal/consensus/auction_test.go
package consensus

import (
	"math/big"
	"testing"
)

// Golden vectors shared with packages/shared/test/auction.test.ts — the two legs of rule A must
// elect the identical winner from the identical inputs.
func e18(n int64) *big.Int {
	return new(big.Int).Mul(big.NewInt(n), big.NewInt(1_000_000_000_000_000_000))
}

func ord(searcher byte, qin, qout *big.Int) ToBOrder {
	return ToBOrder{
		Searcher: addr(searcher),
		PoolID: poolID,
		ZeroForOne: true,
		UseInternal: false,
		QuantityIn: qin,
		QuantityOut: qout,
		ValidForBlock: big.NewInt(100),
	}
}

func TestScoreOrder(t *testing.T) {
	b := ord(0xa2, e18(1), e18(1500))
	want := new(big.Int).Mul(e18(500), q128)
	if got := ScoreOrder(b, price()); got.Cmp(want) != 0 {
		t.Fatalf("score drift: got %s want %s", got, want)
	}
}

func TestElectWinner(t *testing.T) {
	a := ord(0xa1, e18(1), e18(1900)) // surplus 100e18
	b := ord(0xa2, e18(1), e18(1500)) // surplus 500e18 — best
	c := ord(0xa3, e18(1), e18(2100)) // surplus -100e18 — ineligible

	w, ok := ElectWinner([]ToBOrder{a, b, c}, price())
	if !ok || w.Searcher != addr(0xa2) {
		t.Fatalf("winner drift: ok=%v searcher=%x", ok, w.Searcher)
	}
	if _, ok := ElectWinner([]ToBOrder{c}, price()); ok {
		t.Fatalf("non-positive-surplus order should be ineligible")
	}
	if _, ok := ElectWinner(nil, price()); ok {
		t.Fatalf("empty order set should have no winner")
	}
}

func TestElectWinnerTiebreakStable(t *testing.T) {
	d := ord(0xd1, e18(2), e18(3900)) // surplus 100e18
	e := ord(0xe1, e18(2), e18(3900)) // surplus 100e18, same score
	w1, _ := ElectWinner([]ToBOrder{d, e}, price())
	w2, _ := ElectWinner([]ToBOrder{e, d}, price())
	if w1.Searcher != w2.Searcher {
		t.Fatalf("tiebreak not order-independent: %x vs %x", w1.Searcher, w2.Searcher)
	}
}

File: packages/shared/test/auction.test.ts
import { describe, it, expect } from "vitest";
import { electWinner, scoreOrder } from "../src/auction";
import type { ToBOrderT } from "../src/types";

// Golden vectors for the frozen election rule A, shared with the Go port
// (avs/internal/consensus/auction_test.go). Winner = eligible order with the greatest LP surplus at
// the clearing price; ties break on the lower toBStructHash; non-positive-surplus orders are ineligible.
const POOL_ID = "0x1111111111111111111111111111111111111111111111111111111111111111" as const;
const PRICE = 2000n << 128n; // clearing price: 2000 token1 per token0, Q128
const Q128 = 1n << 128n;

// zeroForOne arb: pays quantityIn token0, wants quantityOut token1. surplus = (in*2000 - out) << 128.
function order(searcherLast: string, quantityIn: bigint, quantityOut: bigint): ToBOrderT {
    return {
        searcher: `0x${"0".repeat(38)}${searcherLast}` as `0x${string}`,
        poolId: POOL_ID,
        zeroForOne: true,
        useInternal: false,
        quantityIn,
        quantityOut,
        validForBlock: 100n,
        signature: "0x",
    };
}

const E18 = 1_000_000_000_000_000_000n;
const A = order("a1", E18, 1_900n * E18); // surplus 100e18
const B = order("a2", E18, 1_500n * E18); // surplus 500e18 — best
const C = order("a3", E18, 2_100n * E18); // surplus -100e18 — ineligible

describe("auction rule A (price-weighted net token0)", () => {
    it("scores surplus in token1*Q128 units", () => {
        expect(scoreOrder(B, PRICE)).toBe(500n * E18 * Q128);
        expect(scoreOrder(C, PRICE)).toBe(-100n * E18 * Q128);
    });

    it("elects the greatest-surplus eligible order", () => {
        expect(electWinner([A, B, C], PRICE)?.searcher).toBe(B.searcher);
    });

    it("skips non-positive-surplus orders", () => {
        expect(electWinner([C], PRICE)).toBeNull();
    });

    it("returns null when there are no orders", () => {
        expect(electWinner([], PRICE)).toBeNull();
    });

    it("breaks ties deterministically, independent of input order", () => {
        const d = order("d1", 2n * E18, 3_900n * E18); // surplus 100e18
        const e = order("e1", 2n * E18, 3_900n * E18); // surplus 100e18, same score
        expect(scoreOrder(d, PRICE)).toBe(scoreOrder(e, PRICE));
        expect(electWinner([d, e], PRICE)?.searcher).toBe(electWinner([e, d], PRICE)?.searcher);
    });
});
